gui.js

```diff
@@ -6265,6 +6229,341 @@ ProjectDialogMorph.prototype.fixLayout = function () {
+    this.listField.drawNew = InputFieldMorph.prototype.drawNew;
+    this.listField.drawRectBorder = InputFieldMorph.prototype.drawRectBorder;
+
+    this.listField.action = function (item) {
+        if (isNil(item)) {return; }
+
+        myself.notesText.text = item.description || '';
+        myself.notesText.drawNew();
+        myself.notesField.contents.adjustBounds();
+
+        if (myself.hasCached(item.fileName)) {
+            myself.displayBlocks(item.fileName);
+        } else {
+            myself.showMessage(
+                localize('Loading') + ' ' + localize(item.name)
+            );
+            myself.ide.getURL(
+                myself.ide.resourceURL('libraries', item.fileName),
+                function(libraryXML) {
+                    myself.cacheLibrary(
+                        item.fileName,
+                        libraryXML,
+                        myself.ide.serializer.loadBlocks(libraryXML)
+                    );
+                    myself.displayBlocks(item.fileName);
+                }
+            );
+        }
+    };
+
+    this.body.add(this.listField);
+
+    this.fixLayout();
+};
+
+LibraryImportDialogMorph.prototype.importLibrary = function () {
+    var ide = this.ide,
+        selectedLibrary = this.listField.selected.fileName,
+        libraryName = this.listField.selected.name;
+
+    if (this.hasCached(selectedLibrary)) {
+        ide.droppedText(this.cachedXML(selectedLibrary), libraryName);
+    } else {
+        ide.showMessage(localize('Loading') + ' ' + localize(libraryName));
+        ide.getURL(
+            ide.resourceURL('libraries', selectedLibrary),
+            function(libraryText) {
+                ide.droppedText(libraryText, libraryName);
+            }
+        );
+    }
+
+    this.destroy();
+}
+
+LibraryImportDialogMorph.prototype.displayBlocks = function (libraryKey) {
+    var x, y, blocksList, blockImage, previousCategory, blockContainer,
+        myself = this,
+        padding = 4;
+
+    // Create a copy of the array; we'll remove blocks as we draw them.
+    blocksList = this.cachedBlocks(libraryKey).slice(0);
+
+    if (!blocksList.length) {return; }
+    // populate palette, grouped by categories.
+    this.initializePalette();
+    x = this.palette.left() + padding;
+    y = this.palette.top();
+    SpriteMorph.prototype.categories.forEach(function (category) {
+        blocksList.forEach(function (definition, idx) {
+            if (definition.category !== category) {return; }
+            if (category !== previousCategory) {
+                y += padding;
+            }
+            
+            previousCategory = category;
+            blockImage = definition.templateInstance().fullImage();
+            blockContainer = new Morph();
+            blockContainer.contents = blockImage;
+            blockContainer.image = blockImage;
+            blockContainer.setPosition(new Point(x, y));
+            blockContainer.setExtent(
```

Comment:
Yes, @cycomachead, I believe it's because JS clears <canvas>es when you forcibly resize them. At least, that was the intuition that led me to this solution! :-)